pkg/controller/controller_test.go

```diff
@@ -189,6 +224,158 @@ func newBPWithOutputArtifact() *crv1alpha1.Blueprint {
 	}
 }
 
+func newBPWithDeferPhaseAndErrInDeferPhase() *crv1alpha1.Blueprint {
+	return &crv1alpha1.Blueprint{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "test-blueprint-defer-defererr-",
+		},
+		Actions: map[string]*crv1alpha1.BlueprintAction{
+			"backup": {
+				OutputArtifacts: map[string]crv1alpha1.Artifact{
+					"opArtPhaseOne": {
+						KeyValue: map[string]string{
+							"op": "{{ .Phases.backupPhaseOne.Output.value }}",
+						},
+					},
+					"opArtPhaseTwo": {
+						KeyValue: map[string]string{
+							"op": "{{ .Phases.backupPhaseTwo.Output.value }}",
+						},
+					},
+				},
+				Phases: []crv1alpha1.BlueprintPhase{
+					backupPhase("backupPhaseOne", "mainValue"),
+					backupPhase("backupPhaseTwo", "mainValueTwo"),
+				},
+				DeferPhase: &crv1alpha1.BlueprintPhase{
+					Name: "deferPhase",
+					Func: function.KubeTaskFuncName,
+					Args: map[string]interface{}{
+						"image":     mysqlSidecarImage,
+						"namespace": "default",
+						"command":   []string{"exit", "1"},
+					},
+				},
+			},
+		},
+	}
+}
+
+func newBPWithDeferPhaseAndErrInCorePhase() *crv1alpha1.Blueprint {
+	return &crv1alpha1.Blueprint{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "test-blueprint-defer-coreerr-",
+		},
+		Actions: map[string]*crv1alpha1.BlueprintAction{
+			"backup": {
+				OutputArtifacts: map[string]crv1alpha1.Artifact{
+					"opArtPhaesOne": {
+						KeyValue: map[string]string{
+							"op": "{{ .Phases.backupPhaseOne.Output.value }}",
+						},
+					},
+					"opArtDeferPhase": {
+						KeyValue: map[string]string{
+							"op": "{{ .DeferPhase.Output.value }}",
+						},
+					},
+				},
+				Phases: []crv1alpha1.BlueprintPhase{
+					backupPhase("backupPhaseOne", "mainValue"),
+					{
+						Name: "backupPhaseTwo",
+						Func: function.KubeTaskFuncName,
+						Args: map[string]interface{}{
+							"image":     mysqlSidecarImage,
+							"namespace": "default",
+							"command":   []string{"exit", "1"},
+						},
+					},
+				},
+				DeferPhase: deferPhase(),
+			},
+		},
+	}
+}
+
+func backupPhase(name, outputArtValue string) crv1alpha1.BlueprintPhase {
+	return crv1alpha1.BlueprintPhase{
+		Name: name,
+		Func: function.KubeTaskFuncName,
+		Args: map[string]interface{}{
+			"image":     mysqlSidecarImage,
+			"namespace": "default",
+			"command":   []string{"kando", "output", "value", outputArtValue},
+		},
+	}
+}
+
+func deferPhase() *crv1alpha1.BlueprintPhase {
+	return &crv1alpha1.BlueprintPhase{
+		Name: "deferPhase",
+		Func: function.KubeTaskFuncName,
+		Args: map[string]interface{}{
+			"image":     mysqlSidecarImage,
+			"namespace": "default",
+			"command":   []string{"kando", "output", "value", "deferValue"},
+		},
+	}
+}
+
+func newBPWithDeferPhase() *crv1alpha1.Blueprint {
```

Comment:
Ideally, we should re-use the `NewTestBlueprint()` constructor in the `testutil` package. But the current implementation in `testutil` looks a bit rigid. Maybe in the future we can add better builder interfaces to them.

E.g.,
```go
package testutil

 // without any embedded default action
func NewTestBlueprint() *crv1alpha1.Blueprint

func BlueprintWithAction(blueprint *crv1alpha1.Blueprint, action *crv1alpha1.BlueprintAction) *crv1alpha1.Blueprint

func ActionWithArtifact(action *crv1alpha1.BlueprintAction, artifact *crv1alpha1.Artifact) *crv1alpha1.BlueprintAction

func ActionWithPhase(blueprint *crv1alpha1.Blueprint , phase *crv1alpha1.BlueprintPhase) *crv1alpha1.BlueprintAction 
```

Reply:
yeah, I agree we should track and work on this.